### PR TITLE
BatchNorm: Use resize_output and empty, instead of empty_like

### LINKED ISSUE #62967: BatchNorm2d + SyncBatchNorm incorrect multi gpu behaviour in 1.10.0 (tested working in 1.8.0)
## 🐛 Bug

SyncBatchNorm layers in torch 1.10.0 give different outputs on 2 gpus vs the equivalent BatchNorm layer on a single gpu. This wasn't a problem in torch 1.8.0

## To Reproduce
This code is based on the [PyTorch DDP demo](https://pytorch.org/tutorials/intermediate/ddp_tutorial.html) but the model consists only of a `BatchNorm2d` layer.

```python
import os
import torch
import torch.distributed as dist
import torch.nn as nn
import torch.multiprocessing as mp

from torch.nn.parallel import DistributedDataParallel as DDP
import argparse


def setup(rank, world_size):
    torch.manual_seed(0)
    os.environ["MASTER_ADDR"] = "localhost"
    os.environ["MASTER_PORT"] = "12355"

    # iniialize the process group
    dist.init_process_group("gloo", rank=rank, world_size=world_size)


def cleanup():
    dist.destroy_process_group()


class ToyModel(nn.Module):
    def __init__(self, num_features=10):
        super().__init__()
        self.bn = nn.BatchNorm2d(num_features=num_features)

    def forward(self, x):
        x = x.permute(0, 2, 1, 3)
        x = self.bn(x)
        x = x.permute(0, 2, 1, 3)
        return x


def demo_basic(rank, world_size, convert_sync_batchnorm):
    print(f"Running basic DDP example on rank {rank}.")
    setup(rank, world_size)

    # create model and move it to GPU with id rank
    model = ToyModel(num_features=10)
    # create model and move it to GPU with id rank
    if world_size > 1 and convert_sync_batchnorm:
        print("Converting BatchNorm to SyncBatchNorm")
        model = torch.nn.SyncBatchNorm.convert_sync_batchnorm(model)

    model = model.to(rank)
    if world_size > 1:
        ddp_model = DDP(model, device_ids=[rank])
        model = ddp_model

    inputs = torch.randn(2, 2, 10, 1).to(rank)
    print(f"{rank=}, {inputs[0, 0].squeeze()=}")
    outputs = model(inputs)
    print(f"{rank=}, {outputs[0, 0].squeeze()=}")

    cleanup()


def run_demo(demo_fn, world_size, convert_bn):
    if world_size > 1:
        mp.spawn(demo_fn, args=(world_size, convert_bn), nprocs=world_size, join=True)
    else:
        demo_fn(0, world_size, convert_bn)


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--n_gpus", default=2, type=int, help="Number of gpus")
    parser.add_argument(
        "--convert_bn",
        action="store_true",
        default=False,
        help="Whether to convert BatchNorm layers to SyncBatchNorm",
    )

    args = parser.parse_args()

    print(f"{torch.__version__=}")
    run_demo(demo_basic, args.n_gpus, args.convert_bn)
```

#### 1 - 2 gpus, convert to `SyncBatchNorm`
```
python3 ddp_demo.py --n_gpus 2 --convert_bn
torch.__version__='1.10.0a0+git63955b3'
Running basic DDP example on rank 0.
Running basic DDP example on rank 1.
Converting BatchNorm to SyncBatchNorm
Converting BatchNorm to SyncBatchNorm
rank=1, inputs[0, 0].squeeze()=tensor([-1.1258, -1.1524, -0.2506, -0.4339,  0.8487,  0.6920, -0.3160, -2.1152,
         0.3223, -1.2633], device='cuda:1')
rank=0, inputs[0, 0].squeeze()=tensor([-1.1258, -1.1524, -0.2506, -0.4339,  0.8487,  0.6920, -0.3160, -2.1152,
         0.3223, -1.2633], device='cuda:0')
rank=1, outputs[0, 0].squeeze()=tensor([-1.6701, -0.7058,  0.7740,  0.1321,  0.6478,  0.2026, -0.1025,  1.0656,
        -1.3900,  0.9416], device='cuda:1', grad_fn=<SqueezeBackward0>)
rank=0, outputs[0, 0].squeeze()=tensor([-1.6701, -0.7058,  0.7740,  0.1321,  0.6478,  0.2026, -0.1025,  1.0656,
        -1.3900,  0.9416], device='cuda:0', grad_fn=<SqueezeBackward0>)
```

#### 2 - 2 gpus, do not convert to `SyncBatchNorm`
```
python3 ddp_demo.py --n_gpus 2
torch.__version__='1.10.0a0+git63955b3'
Running basic DDP example on rank 0.
Running basic DDP example on rank 1.
rank=1, inputs[0, 0].squeeze()=tensor([-1.1258, -1.1524, -0.2506, -0.4339,  0.8487,  0.6920, -0.3160, -2.1152,
         0.3223, -1.2633], device='cuda:1')
rank=0, inputs[0, 0].squeeze()=tensor([-1.1258, -1.1524, -0.2506, -0.4339,  0.8487,  0.6920, -0.3160, -2.1152,
         0.3223, -1.2633], device='cuda:0')
rank=1, outputs[0, 0].squeeze()=tensor([-1.6701, -0.7582, -0.7058, -1.2356,  0.7740,  1.4852,  0.1321, -1.1821,
         0.6478, -1.1941], device='cuda:1', grad_fn=<SqueezeBackward0>)
rank=0, outputs[0, 0].squeeze()=tensor([-1.6701, -0.7582, -0.7058, -1.2356,  0.7740,  1.4852,  0.1321, -1.1821,
         0.6478, -1.1941], device='cuda:0', grad_fn=<SqueezeBackward0>)
```

#### 3 - 1 gpu
```
python3 ddp_demo.py --n_gpus 1
torch.__version__='1.10.0a0+git63955b3'
Running basic DDP example on rank 0.
False
rank=0, inputs[0, 0].squeeze()=tensor([-1.1258, -1.1524, -0.2506, -0.4339,  0.8487,  0.6920, -0.3160, -2.1152,
         0.3223, -1.2633], device='cuda:0')
rank=0, outputs[0, 0].squeeze()=tensor([-1.6701, -0.7582, -0.7058, -1.2356,  0.7740,  1.4852,  0.1321, -1.1821,
         0.6478, -1.1941], device='cuda:0', grad_fn=<SqueezeBackward0>)
```

1 != 2 ❌
2 == 3 ✅

#### 1 - 2 gpus, convert to `SyncBatchNorm`
```
python3 ddp_demo.py --n_gpus 2 --convert_bn
torch.__version__='1.8.0a0+618cf40'
Running basic DDP example on rank 0.
Running basic DDP example on rank 1.
Converting BatchNorm to SyncBatchNorm
Converting BatchNorm to SyncBatchNorm
rank=1, inputs[0, 0].squeeze()=tensor([-1.1258, -1.1524, -0.2506, -0.4339,  0.8487,  0.6920, -0.3160, -2.1152,
         0.3223, -1.2633], device='cuda:1')
rank=0, inputs[0, 0].squeeze()=tensor([-1.1258, -1.1524, -0.2506, -0.4339,  0.8487,  0.6920, -0.3160, -2.1152,
         0.3223, -1.2633], device='cuda:0')
rank=0, outputs[0, 0].squeeze()=tensor([-1.6701, -0.7582, -0.7058, -1.2356,  0.7740,  1.4852,  0.1321, -1.1821,
         0.6478, -1.1941], device='cuda:0', grad_fn=<SqueezeBackward0>)
rank=1, outputs[0, 0].squeeze()=tensor([-1.6701, -0.7582, -0.7058, -1.2356,  0.7740,  1.4852,  0.1321, -1.1821,
         0.6478, -1.1941], device='cuda:1', grad_fn=<SqueezeBackward0>)
```

#### 2 - 2 gpus, do not convert to `SyncBatchNorm`
```
python3 ddp_demo.py --n_gpus 2
torch.__version__='1.8.0a0+618cf40'
Running basic DDP example on rank 0.
Running basic DDP example on rank 1.
rank=0, inputs[0, 0].squeeze()=tensor([-1.1258, -1.1524, -0.2506, -0.4339,  0.8487,  0.6920, -0.3160, -2.1152,
         0.3223, -1.2633], device='cuda:0')
rank=1, inputs[0, 0].squeeze()=tensor([-1.1258, -1.1524, -0.2506, -0.4339,  0.8487,  0.6920, -0.3160, -2.1152,
         0.3223, -1.2633], device='cuda:1')
rank=1, outputs[0, 0].squeeze()=tensor([-1.6701, -0.7582, -0.7058, -1.2356,  0.7740,  1.4852,  0.1321, -1.1821,
         0.6478, -1.1941], device='cuda:1', grad_fn=<SqueezeBackward0>)
rank=0, outputs[0, 0].squeeze()=tensor([-1.6701, -0.7582, -0.7058, -1.2356,  0.7740,  1.4852,  0.1321, -1.1821,
         0.6478, -1.1941], device='cuda:0', grad_fn=<SqueezeBackward0>)
```

#### 3 - 1 gpu
```
python3 ddp_demo.py --n_gpus 1
torch.__version__='1.8.0a0+618cf40'
Running basic DDP example on rank 0.
rank=0, inputs[0, 0].squeeze()=tensor([-1.1258, -1.1524, -0.2506, -0.4339,  0.8487,  0.6920, -0.3160, -2.1152,
         0.3223, -1.2633], device='cuda:0')
rank=0, outputs[0, 0].squeeze()=tensor([-1.6701, -0.7582, -0.7058, -1.2356,  0.7740,  1.4852,  0.1321, -1.1821,
         0.6478, -1.1941], device='cuda:0', grad_fn=<SqueezeBackward0>)
```

1 == 2 == 3 ✅

## Expected behavior

I expect the behaviour in torch 1.10.0 to mirror that in 1.8.0 where options 1, 2 and 3 all produce the same output.

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

### Torch 1.10.0a0+git63955b3 Env info
```bash
Collecting environment information...
PyTorch version: 1.10.0a0+git63955b3
Is debug build: False
CUDA used to build PyTorch: 11.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.21.1
Libc version: glibc-2.27

Python version: 3.8.11 (default, Jul  3 2021, 17:53:42)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-65-generic-x86_64-with-glibc2.27
Is CUDA available: False
CUDA runtime version: 11.2.152
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.1.1
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.21.1
[pip3] pytorch-memlab==0.2.3
[pip3] torch==1.10.0a0+git63955b3
[pip3] torch-tb-profiler==0.2.1
[pip3] torchaudio==0.9.0a0-jit-kaldi-fa498ad
[pip3] torchpq==0.1.8.3
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.10.0

```

### Torch 1.8.0a0+618cf40 Env info
```bash
Collecting environment information...
PyTorch version: 1.8.0a0+618cf40
Is debug build: False
CUDA used to build PyTorch: 11.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.21.1
Libc version: glibc-2.27

Python version: 3.8.11 (default, Jul  3 2021, 17:53:42)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-65-generic-x86_64-with-glibc2.27
Is CUDA available: False
CUDA runtime version: 11.2.152
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.1.1
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.21.1
[pip3] pytorch-memlab==0.2.3
[pip3] torch==1.8.0a0+618cf40
[pip3] torchaudio==0.9.0a0-jit-kaldi-fa498ad
[pip3] torchpq==0.1.8.3
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.10.0
[conda] Could not collect
``` 

cc @albanD @mruberry @jbschlosser

### PR DESCRIPTION
**Pull request by peterbell10**
Fixes #62967

This lets each of the three implementations choose which memory format
to use for the output, meaning channels_last can be used in more cases.

### CODE REVIEW

File: aten/src/ATen/native/cuda/Normalization.cu [hunk 1]
```diff
@@ -423,7 +429,7 @@ std::tuple<Tensor&, Tensor&, Tensor&> batch_norm_cuda_out(const Tensor& self, co
 }
 
 std::tuple<Tensor, Tensor, Tensor> batch_norm_cuda(const Tensor& self, const c10::optional<Tensor>& weight_opt, const c10::optional<Tensor>& bias_opt, const c10::optional<Tensor>& running_mean_opt, const c10::optional<Tensor>& running_var_opt, bool train, double momentum, double epsilon) {
-  auto output = at::empty_like(self, self.suggest_memory_format());
+  auto output = at::empty({0}, self.options());
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Normalization.cu:432]: I think it's better to just use `empty_like` here, without specifying memory format. `empty_like` propagates permutation, so most regular tensors (including the one from the issue) will have strides same as `self`, and will pass `is_contiguous` check. Then you wouldn't need special logic in `batch_norm_elementwise`. On the other hand, `resize_impl_cuda_(out.unsafeGetTensorImpl(), self.sizes(), self.strides());` that you have now can allocate non-memory-dense or self-overlapping tensor, neither of which is good.

File: aten/src/ATen/native/cuda/Normalization.cuh [hunk 2]
```diff
@@ -694,6 +694,8 @@ void batch_norm_stats_cuda_template(
   Tensor dummy_var_;
   auto input_reshaped = input_.reshape({input_.size(0), input_.size(1), -1}); // internally we merge the feature dimensions
 
+  resize_output(out_mean, {n_input});
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Normalization.cuh:697]: why do we need this?

File: aten/src/ATen/native/cuda/Normalization.cu (hunk 1 above)
Comment (peterbell10) [reply aten/src/ATen/native/cuda/Normalization.cu]: Okay, returned to call `empty_like` but we still need `resize_impl_cuda_` for the out-argument overload. Note that it can't create non-dense tensors because the channels last implementation only supports dense tensors as input.

File: aten/src/ATen/native/cuda/Normalization.cuh (hunk 2 above)
Comment (peterbell10) [reply aten/src/ATen/native/cuda/Normalization.cuh:697]: Not strictly necessary for this change, but I noticed the out-argument variants didn't resize outputs correctly.

**Review by ngimel — APPROVED** [review]
